api/v1beta1/machinedrainrules_types.go

```diff
@@ -0,0 +1,213 @@
+	// Pods with higher order are drained after Pods with lower order.
+	// order can only be set if behavior is set to "Drain".
+	// If order is not set, 0 will be used.
+	// Valid values for order are from -2147483648 to 2147483647 (inclusive).
+	// +optional
+	Order *int32 `json:"order,omitempty"`
+}
+
+type MachineDrainRuleMachineSelector struct { //nolint:revive // Intentionally not adding a godoc comment as it would show up additionally to the field comment in the CRD
+	// selector is a label selector which selects Machines by their labels.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects all Machines.
+	//
+	// If clusterSelector is also set, then the selector as a whole selects
+	// Machines matching selector belonging to Clusters selected by clusterSelector.
+	// If clusterSelector is not set, it selects all Machines matching selector in
+	// all Clusters.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// clusterSelector is a label selector which selects Machines by the labels of
+	// their Clusters.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects Machines of all Clusters.
+	//
+	// If selector is also set, then the selector as a whole selects
+	// Machines matching selector belonging to Clusters selected by clusterSelector.
+	// If selector is not set, it selects all Machines belonging to Clusters
+	// selected by clusterSelector.
+	// +optional
+	ClusterSelector *metav1.LabelSelector `json:"clusterSelector,omitempty"`
+}
+
+type MachineDrainRulePodSelector struct { //nolint:revive // Intentionally not adding a godoc comment as it would show up additionally to the field comment in the CRD
+	// selector is a label selector which selects Pods by their labels.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects all Pods.
+	//
+	// If namespaceSelector is also set, then the selector as a whole selects
+	// Pods matching selector in Namespaces selected by namespaceSelector.
+	// If namespaceSelector is not set, it selects all Pods matching selector in
+	// all Namespaces.
+	// +optional
+	Selector *metav1.LabelSelector `json:"selector,omitempty"`
+
+	// namespaceSelector is a label selector which selects Pods by the labels of
+	// their Namespaces.
+	// This field follows standard label selector semantics; if not present or
+	// empty, it selects Pods of all Namespaces.
+	//
+	// If selector is also set, then the selector as a whole selects
+	// Pods matching selector in Namespaces selected by namespaceSelector.
+	// If selector is not set, it selects all Pods in Namespaces selected by
+	// namespaceSelector.
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machinedrainrules,shortName=mdr,scope=Namespaced,categories=cluster-api
```

Comment:
Shortnames are generally not a good idea. Officially sig-cli only supports fully qualified resource names, as short names have high coincidence rates for conflicts.

Shortnames should be reserved ideally for resources in use regularly by the majority of end users, perhaps we could argue that a management cluster end user is likely to use this frequently, but, I'd still lean more to reserving shortnames for workload rather than infrastructure related CRDs

Reply:
Dropped